### PR TITLE
THREESCALE-7919: Added content for the DestinationRule custom resource

### PR DESCRIPTION
**Pull request by dfennessy**
Work done for Jira Issue [THREESCALE-7919](https://issues.redhat.com/browse/THREESCALE-7919)

**Reviewers**

* @JStickler
* @rahulanand16nov (3scale SME)
* @azgabur (3scale QE)

**Preview:** https://47204--docspreview.netlify.app/openshift-enterprise/latest/service_mesh/v2x/ossm-threescale-webassembly-module.html#ossm-threescale-applying-external-service-entry-objects_ossm-threescale-webassembly-module

~~http://file.rdu.redhat.com/dfenness/THREESCALE-7919/service_mesh/v2x/ossm-threescale-webassembly-module.html#ossm-threescale-applying-external-service-entry-objects_ossm-threescale-webassembly-module~~

### CODE REVIEW

**Comment by dfennessy** [comment]
@JStickler @rahulanand16nov

The preview is in the description.

**Review by rahulanand16nov — APPROVED** [review]

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (JStickler) [inline modules/ossm-threescale-applying-external-service-entry-objects.adoc:9]: When we merged the service mesh attributes file into the OpenShift common-attributes file, we changed this attribute name.  Use SMProductName.

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (JStickler) [inline modules/ossm-threescale-applying-external-service-entry-objects.adoc:120]: "To do this, change the location of these services in the CRs."  Too many vague pronoun references here, this is going to  confuse anyone who doesn't know the product extremely well.   Name what task you're doing, the services whose location needs to be changed, and which specific CR needs to be updated.

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (JStickler) [inline modules/ossm-threescale-applying-external-service-entry-objects.adoc:87]: Check the preview, there are four bullets, but it looks like two of them are repeats?  Possibly a copy/paste error?

**Review by JStickler — COMMENTED** [review]
I still find the order of this procedure to be really confusing.  

(Also, glad you figured out how to add a preview!)

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (azgabur) [inline modules/ossm-threescale-applying-external-service-entry-objects.adoc:15]: Both procedure steps are very similar and have only difference is in words "backend" in first step and "system" in second step. I would make these words **bold**  or otherwise distinguish them to stand out a little.

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (dfennessy) [reply modules/ossm-threescale-applying-external-service-entry-objects.adoc:87]: These are different for system and backend. For example, the commands in step 1 vs step 3. The same applies to step 2 and step 4.

1. `$ oc apply -f <service-entry-threescale-saas-backend.yml>`
`$ oc apply -f <destination-rule-threescale-saas-backend.yml>`

3. $ oc apply -f <service-entry-threescale-saas-system.yml>
$ oc apply -f <destination-rule-threescale-saas-system.yml>

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (dfennessy) [reply modules/ossm-threescale-applying-external-service-entry-objects.adoc:15]: Thanks @azgabur That's a good idea based on the previous feedback from Julie.

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (JStickler) [reply modules/ossm-threescale-applying-external-service-entry-objects.adoc:87]: There is no text that distinguishes between the two sets of bullets.  It's pretty much impossible to tell that one set of commands is intended for the  system and one for the backend. Is the text "3scale Hosted backend and system" supposed to be a heading (no space between the period and the text in the markup)?  Or is it supposed to be a procedure step (needs a space between the period and the text in the markup)?  It might be clearer to actually make this two separate steps, one for the system and one for the backend.

Comment (dfennessy) [reply modules/ossm-threescale-applying-external-service-entry-objects.adoc:87]: Sorry about the lack of clarity. The text with no space between the period and the text creates a title or heading for that which is under it if I'm not mistaken. It is not a procedure step, which as we know require a space between the period and text.

In any case, I will make this clearer.

I have an issue with this PR. For the last commit, I am getting this message "This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository." I do not know why it happened, but I think made changes based on the last comments from you and Alex, however none of them showed up in the commit.

I am going to have to close this PR and create a new one with the updates. My apologies for the inconvenience (if any) this causes. I think I've everything I need to progress with the PR now, including feedback from yourself and Alex. Hopefully nearly there. This has been delayed long enough.

Comment (dfennessy) [reply modules/ossm-threescale-applying-external-service-entry-objects.adoc]: **Update:** My latest commit seems to have end up in this PR, although I started from scratch locally, so I guess I'll keep it open. Let's see if all the checks pass.

**Comment by dfennessy** [comment]
@azgabur I've made the update we discussed on Slack: https://47204--docspreview.netlify.app/openshift-enterprise/latest/service_mesh/v2x/ossm-threescale-webassembly-module.html#ossm-threescale-applying-external-service-entry-objects_ossm-threescale-webassembly-module

**Review by azgabur — APPROVED** [review]
Looks good! I give it LGTM

**Comment by dfennessy** [comment]
> Looks good! I give it LGTM

Awesome! Thanks @azgabur 

@JStickler Hopefully we can now merge this PR. Cheers :slightly_smiling_face:

**Comment by dfennessy** [comment]
> > Looks good! I give it LGTM
> 
> Awesome! Thanks @azgabur
> 
> @JStickler Hopefully we can now merge this PR. Cheers slightly_smiling_face

cherry-pick 4.6 to 4.11

**Review by JStickler — CHANGES_REQUESTED** [review]
Please reorganize and reformat so that the commands are grouped with the files they apply to.

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (dfennessy) [inline modules/ossm-threescale-applying-external-service-entry-objects.adoc:122]: @rahulanand16nov  Is this alternative way to deploy an in-mesh 3scale service necessary to document here or would it be better to remove this line so as not to cause confusion?

**Comment by dfennessy** [comment]
FYI @azgabur 

https://47204--docspreview.netlify.app/openshift-enterprise/latest/service_mesh/v2x/ossm-threescale-webassembly-module.html#ossm-threescale-applying-external-service-entry-objects_ossm-threescale-webassembly-module

**Review by azgabur — APPROVED** [review]
LGTM!

**Comment by dfennessy** [comment]
/label peer-review-needed

This was originally under review by Julie Stickler. I made updates based on her last peer review, so it needs at least on more review. Cheers.

3scale eng/QE have approved. 

Thanks in advance for your support :slightly_smiling_face: 

Once approved, please merge and cherry-pick: 4.6 to 4.11

**Review by tmalove — COMMENTED** [review]
/remove-label peer-review-needed
/label peer-review-in-progress

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (tmalove) [inline modules/ossm-threescale-applying-external-service-entry-objects.adoc:9]:
```suggestion
To have the `threescale-wasm-auth` module authorize requests against 3scale, the module must have access to 3scale services. You can do this within {SMProductName} by applying an external `ServiceEntry` object and a corresponding `DestinationRule` object for TLS configuration to use the HTTPS protocol.
```
Suggestion to change to 'do' for simple language.

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (tmalove) [inline modules/ossm-threescale-applying-external-service-entry-objects.adoc:53]:
```suggestion
.. Save the *backend* objects to one or more files, and use the following syntax:
```
Delete for minimalism.

**Review by tmalove — COMMENTED** [review]
/remove-label peer-review-in-progress
/label peer-review-done

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (dfennessy) [reply modules/ossm-threescale-applying-external-service-entry-objects.adoc:9]: Thank you @tmalove :slightly_smiling_face:

**Comment by jboxman-rh** [comment]
@dfennessy, is that for a re-review? Thanks!

**Comment by dfennessy** [comment]
> @dfennessy, is that for a re-review? Thanks!

Yes, my apologies. I thought @tmalove would sign off on the latest updates from her suggestions.

Thanks @jboxman-rh

**Comment by dfennessy** [comment]
> > Once approved, please merge and cherry-pick: 4.6 to 4.11
> 
> @dfennessy hi! Is there a specific reason why this should not also be cherry picked to 4.12? Thanks!

Hi Paul,
No reason. Please cherry-pick to 4.12 also. Cheers.

**Review by pneedle-rh — COMMENTED** [review]
@dfennessy I have completed my review of this PR. Please let me know if you have any questions about my suggestions. Thanks!

**Comment by pneedle-rh** [comment]
> Hi Paul, No reason. Please cherry-pick to 4.12 also. Cheers.

Ok, thank you for confirming that. I have subsequently added the `enterprise-4.12` label.

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (dfennessy) [reply modules/ossm-threescale-applying-external-service-entry-objects.adoc:15]: The capitalization in "3scale Hosted" is part of our standards in the 3scale documentation. 

For example, the standard states that: _We capitalize the occurrence of the product name and the deployment type_. So when hosted comes after 3scale, we capitalize.

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (dfennessy) [reply modules/ossm-threescale-applying-external-service-entry-objects.adoc:17]: I was previous told that the back-ticks are fine for block title syntax heading (titles), however I am happy to remove them.

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (dfennessy) [reply modules/ossm-threescale-applying-external-service-entry-objects.adoc:50]: Thanks, this makes sense.

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (dfennessy) [reply modules/ossm-threescale-applying-external-service-entry-objects.adoc:57]: Cheers. I will make this update.

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (dfennessy) [reply modules/ossm-threescale-applying-external-service-entry-objects.adoc:15]: Thank you. I will make this update.

**Comment by dfennessy** [comment]
> @dfennessy I have completed my review of this PR. Please let me know if you have any questions about my suggestions. Thanks!

@pneedle-rh  Thanks for thorough review. I've made updates based on your suggestions.

**Comment by dfennessy** [comment]
The PR has been approved by 3scale SME and QE.

I need a peer reviewer LGTM so that I can are the label for `merge-review-needed`.

Cheers

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (pneedle-rh) [inline modules/ossm-threescale-applying-external-service-entry-objects.adoc:16]: I recommend using back-ticks for the file name references, instead of italicising them. This is consistent with other examples in the docs.

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (pneedle-rh) [inline modules/ossm-threescale-applying-external-service-entry-objects.adoc:35]: I recommend using back-ticks for the file name references, instead of italicising them. This is consistent with other examples in the docs.

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (pneedle-rh) [inline modules/ossm-threescale-applying-external-service-entry-objects.adoc:99]: I recommend using back-ticks for the file name references, instead of italicising them. This is consistent with other examples in the docs.

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (pneedle-rh) [inline modules/ossm-threescale-applying-external-service-entry-objects.adoc:59]: This command as far as I can tell would require a previously saved `service-entry-threescale-saas-backend.yml` file and it just applies the configuration in much the same way as the preceding command. You could remove this alternative step entirely to minimise the procedure.

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (pneedle-rh) [inline modules/ossm-threescale-applying-external-service-entry-objects.adoc:72]: This command as far as I can tell would require a previously saved `destination-rule-threescale-saas-backend.yml` file and it just applies the configuration in much the same way as the preceding command. You could remove this alternative step entirely to minimise the procedure.

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (pneedle-rh) [inline modules/ossm-threescale-applying-external-service-entry-objects.adoc:123]: This command as far as I can tell would require a previously saved `service-entry-threescale-saas-system.yml` file and it just applies the configuration in much the same way as the preceding command. You could remove this alternative step entirely to minimise the procedure.

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (pneedle-rh) [inline modules/ossm-threescale-applying-external-service-entry-objects.adoc:136]: This command as far as I can tell would require a previously saved `destination-rule-threescale-saas-system.yml` file and it just applies the configuration in much the same way as the preceding command. You could remove this alternative step entirely to minimise the procedure.

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (pneedle-rh) [inline modules/ossm-threescale-applying-external-service-entry-objects.adoc:130]: You could make this a substep (`..`) rather than a step (`.`), because it is part of the first overall step to apply the CRs.

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (pneedle-rh) [inline modules/ossm-threescale-applying-external-service-entry-objects.adoc:134]: The `service-entry-threescale-saas-backend.yml` reference in this command is passed literally to the command line and it is not a variable. This is because the preceding substep requested that the file is saved explicitly using this file name. You could remove the parenthesis:

> $ oc apply -f destination-rule-threescale-saas-system.yml

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (pneedle-rh) [inline modules/ossm-threescale-applying-external-service-entry-objects.adoc:120]: The `service-entry-threescale-saas-backend.yml` reference in this command is passed literally to the command line and it is not a variable. This is because the preceding substep requested that the file is saved explicitly using this file name. You could remove the parenthesis:

> $ oc apply -f service-entry-threescale-saas-system.yml

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (pneedle-rh) [inline modules/ossm-threescale-applying-external-service-entry-objects.adoc:70]: The `service-entry-threescale-saas-backend.yml` reference in this command is passed literally to the command line and it is not a variable. This is because the preceding substep requested that the file is saved explicitly using this file name. You could remove the parenthesis:

> $ oc apply -f destination-rule-threescale-saas-backend.yml

**Review by pneedle-rh — COMMENTED** [review]
@dfennessy I have added a few additional comments, following the updates relating to the structure of the procedural steps. Please let me know if you have any questions about my additional suggestions. Thanks!

**Comment by dfennessy** [comment]
> @dfennessy I have added a few additional comments, following the updates relating to the structure of the procedural steps. Please let me know if you have any questions about my additional suggestions. Thanks!

@pneedle-rh Thanks. Updates made.

**Comment by pneedle-rh** [comment]
LGTM! Thank you for addressing those suggestions.

**Comment by dfennessy** [comment]
Peer, SME, and QE review complete.

/label merge-review-needed

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (stevsmit) [inline modules/ossm-threescale-applying-external-service-entry-objects.adoc:35]: I cannot find this in the preview.

**Comment by stevsmit** [comment]
@dfennessy There is content in this PR that I could not find in the preview.

File: modules/ossm-threescale-applying-external-service-entry-objects.adoc
Comment (dfennessy) [reply modules/ossm-threescale-applying-external-service-entry-objects.adoc:35]: Ok! Thanks, Steven.

**Comment by dfennessy** [comment]
> /label merge-review-needed

@stevsmit Please see the link in the description. The previous link was from when we had to generate builds manually. 

Cheers.

/label merge-review-needed

**Pull request merged**
